src/grav/client.py

```diff
@@ -0,0 +1,119 @@
+
+    1) test the daemon
+    2) show my development ability
+
+I've chosen to write a minimal docker client.
+"""
+
+
+class FriendlyHTTPResponse(HTTPResponse):
+    # I really like requests .json() method on response objects.
+    # this class is solely to enable that -- wdella 2019-10
+    def __init__(self, *args, **kwargs):
+        super(FriendlyHTTPResponse, self).__init__(*args, **kwargs)
+
+    def json(self):
+        """Reads the response body, parses it as json, and returns the result."""
+        return json.loads(self.read())
+
+
+class SocketHTTPConnection(HTTPConnection):
+    """
+    Wrapper to allow httplib to be used with a file socket.
+
+    HTTPConnection typically expects to be given a http url.
+    This is the minimal glue to allow it to be used with a
+    file path to a socket instead.
+    """
+    def __init__(self, base_url, socket_path, timeout=10):
+        super(SocketHTTPConnection, self).__init__(
+            'localhost', timeout=timeout
+        )
+        self.socket_path = socket_path
+        self.timeout = timeout
+
+    def connect(self):
+        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        sock.settimeout(self.timeout)
+        sock.connect(self.socket_path)
+        self.sock = sock
+
+    def response_class(self, sock, *args, **kwargs):
+        return FriendlyHTTPResponse(sock, *args, **kwargs)
+
+
+class DockerClient:
+    """
+    A minimal docker client that can GET & POST to a socket.
+    """
+    def __init__(self, target):
+        self._target = target
+        # With more time, I'd do  connection pooling instead of
+        # a new connection for each request
+        self.conn = SocketHTTPConnection("/", self._target)
+        self.conn.connect()
+
+    def get(self, path, headers={}):
+        return self.request("GET", path, headers=headers)
+
+    def post(self, path, body=None, headers={}):
+        return self.request("POST", path, body=body, headers=headers)
+
+    def request(self, verb, path, body=None, headers={}):
+        if type(body) == dict:  # automatic json conversion
+            body = json.dumps(body)
+        self.conn.request(verb, path, body=body, headers=headers)
+        resp = self.conn.getresponse()
+        return resp
+
+    # make sure not to  leak connections, close when the client falls out of scope
+    def __del__(self):
+        self.conn.close()
+
+
+def demux_logs(log_stream):
+    """
+    Docker uses a multiplexed format for returning logs.
+
+    Split a single byte stream into two vanilla strings.
+
+    returns stdout, stderr
+    """
+    if type(log_stream) == bytes:
+        log_stream = BytesIO(log_stream)
+
+    stdout = stderr = b''
+    while True:
+        header = log_stream.read(8)
+        if not header:  # EOF
+            break
+        stream, content_len = unpack('>BxxxL', header)
+        if stream == 1:  # stdout
+            stdout += log_stream.read(content_len)
```

Comment:
performance/style nit - this would be faster if this used StringIO instead of concatenation especially for large byte streams.

Reply:
I can change this if you feel strongly about it.  Certainly stringbuilders are well recognized pattern for this.

The reason I chose not to bother was: In the challenge doc, I was told:

> Do not implement a system that scales or is highly performing. What scheduling performance improvements would you add in the future?

Honestly, the client code is rife with other perf/determinism issues too (e.g. no connection pooling or retries).

Reply:
The way I set the demux API up also requires that all of stdout & stderr be in memory at the same time at the end of the function.

If we wanted to have a true streaming API, I'd need to yield a tuple similar to https://github.com/docker/docker-py/blob/b72fb1e571d1019f2533c931bc2f68a7d4b71db5/docker/utils/socket.py#L88-L106 or have something like subprocess's api for stdout and stderr.  Either one of those seems like an improvement over this, but feels like overengineering for the scope we're working in hwere.

Reply:
Makes sense, however what we meant is that we don't expect anyone to spend additional efforts on building a scalable solution compared to what would be required to the baseline MVP, like adding caching layer or connection pooling. However in this case the cost of faster solution == cost of slower and expensive solution, (one to one line), that is why I'm making this comment.

Reply:
I think 07d71b3 covers what you're asking for.

If you were talking about something heavier (e.g. fixing the flattening at the end of `demux_logs`) I'd like to discuss what sort of API you think is appropriate for this situation.

Reply:
Without any further comment from you @javins klizhentas, I'm going to assume 07d71b3 addresses this concern and mark this issue resolved.

If there was something else you were suggesting, reopen and I'll get it incorporated into a future PR.